File: Assignment/Chapter 1/Student Grade Calculator/Kotlin Version/app/src/main/java/com/romualdsigning/studentgradecalc/data/export/ReportExporterFactory.kt
package com.romualdsigning.studentgradecalc.data.export

import com.romualdsigning.studentgradecalc.data.io.WorkbookExportService
import com.romualdsigning.studentgradecalc.domain.model.ExportFormat
import com.romualdsigning.studentgradecalc.domain.service.ReportExporter

class ReportExporterFactory(
    private val excelExporter: ReportExporter = WorkbookExportService(),
    private val csvExporter: ReportExporter = CsvReportExporter(),
    private val jsonExporter: ReportExporter = JsonReportExporter(),
    private val pdfExporter: ReportExporter = PdfReportExporter(),
    private val wordExporter: ReportExporter = WordReportExporter(),
) {
    fun create(format: ExportFormat): ReportExporter =
        when (format) {
            ExportFormat.EXCEL -> excelExporter
            ExportFormat.CSV -> csvExporter
            ExportFormat.JSON -> jsonExporter
            ExportFormat.PDF -> pdfExporter
            ExportFormat.WORD -> wordExporter
        }
}

File: Assignment/Chapter 1/Student Grade Calculator/Dart Version/lib/core/services/export/report_exporter_factory.dart
import '../../contracts/report_exporter.dart';
import '../../models/export_format.dart';
import '../workbook_export_service.dart';
import 'csv_report_exporter.dart';
import 'json_report_exporter.dart';
import 'pdf_report_exporter.dart';
import 'word_report_exporter.dart';

class ReportExporterFactory {
  const ReportExporterFactory({
    this.excelExporter = const WorkbookExportService(),
    this.csvExporter = const CsvReportExporter(),
    this.jsonExporter = const JsonReportExporter(),
    this.pdfExporter = const PdfReportExporter(),
    this.wordExporter = const WordReportExporter(),
  });

  final ReportExporter excelExporter;
  final ReportExporter csvExporter;
  final ReportExporter jsonExporter;
  final ReportExporter pdfExporter;
  final ReportExporter wordExporter;

  ReportExporter create(ExportFormat format) => switch (format) {
    ExportFormat.excel => excelExporter,
    ExportFormat.csv => csvExporter,
    ExportFormat.json => jsonExporter,
    ExportFormat.pdf => pdfExporter,
    ExportFormat.word => wordExporter,
  };
}
